### PR TITLE
✨ ERC1155

### PR DESCRIPTION
**Pull request by abigger87**
Adapted from [0xsequence ERC1155 Impl](https://github.com/0xsequence/erc-1155)

### File Changes

> src/tokens
- [x] `ERC1155.sol`
- [x] `ERC165.sol`
- [x] `ERC1155TokenReceiver.sol`

### CODE REVIEW

**Comment by transmissions11** [comment]
can u remove all the weird vscode files n stuff in here pls

**Comment by transmissions11** [comment]
we also dont do interfaces in solmate, the contract is the interface

**Comment by transmissions11** [comment]
can put `ERC1155TokenReceiver` in the ERC1155 contract

File: src/tokens/ERC1155.sol [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+                    Internal Transfer Functions
+  //////////////////////////////////////////////////////////////*/
+
+  /// @notice Transfers amount amount of an _id from the _from address to the _to address specified
+  /// @param _from    Source address
+  /// @param _to      Target address
+  /// @param _id      ID of the token type
+  /// @param _amount  Transfered amount
+  function _safeTransferFrom(address _from, address _to, uint256 _id, uint256 _amount) internal {
+    balances[_from][_id] = balances[_from][_id].sub(_amount); // Subtract amount from
```
Comment (transmissions11) [inline src/tokens/ERC1155.sol:70]: where are these safemath funcs coming from? this is 0.8

File: src/tokens/ERC165.sol [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import {IERC165} from "../interfaces/IERC165.sol";
+
+/// @notice Modern and gas efficient ERC165 implementation.
+/// @author Modified from 0xsequence (https://github.com/0xsequence/erc-1155)
+abstract contract ERC165 is IERC165 {
+
+  /// @notice Query if a contract implements an interface
+  /// @param _interfaceID The interface identifier, as specified in ERC-165
+  /// @return `true` if the contract implements `_interfaceID`
+  function supportsInterface(bytes4 _interfaceID) virtual override public pure returns (bool) {
```
Comment (transmissions11) [inline src/tokens/ERC165.sol:13]: just put this in the erc1155 no seperate contract bls

File: src/tokens/ERC1155.sol [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  }
+
+  /*///////////////////////////////////////////////////////////////
+                        Operator Functions
+  //////////////////////////////////////////////////////////////*/
+
+  /// @notice Enable or disable approval for a third party ("operator") to manage all of caller's tokens
+  /// @param _operator  Address to add to the set of authorized operators
+  /// @param _approved  True if the operator is approved, false to revoke approval
+  function setApprovalForAll(address _operator, bool _approved) external override {
```
Comment (transmissions11) [inline src/tokens/ERC1155.sol:134]: we dont use underlines for variable names :)

Comment (abigger87) [reply src/tokens/ERC1155.sol:134]: okie

File: src/tokens/ERC1155.sol [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+  /// @param _to       Target addresses
+  /// @param _ids      IDs of each token type
+  /// @param _amounts  Transfer amounts per token type
+  /// @param _gasLimit The gas limit
+  /// @param _data     Data to call with
+  function _callonERC1155BatchReceived(address _from, address _to, uint256[] memory _ids, uint256[] memory _amounts, uint256 _gasLimit, bytes memory _data) internal {
+    // Pass data if recipient is contract
+    if (_to.code.length > 0) {
+      bytes4 retval = IERC1155TokenReceiver(_to).onERC1155BatchReceived{gas: _gasLimit}(msg.sender, _from, _ids, _amounts, _data);
+      require(retval == ERC1155_BATCH_RECEIVED_VALUE, "ERC1155#_callonERC1155BatchReceived: INVALID_ON_RECEIVE_MESSAGE");
```
Comment (transmissions11) [inline src/tokens/ERC1155.sol:123]: remove all the prefix stuff from errs pls, just INVALID_ON_RECEIVE_MESSAGE or whatevs

File: src/tokens/ERC1155.sol (hunk 3 above)
Comment (abigger87) [reply src/tokens/ERC1155.sol:134]: sry shoulda drafted pr

Comment (abigger87) [reply src/tokens/ERC1155.sol:134]: will conform stronger rn

**Comment by transmissions11** [comment]
also no natspec or comments unless behavior super confusing pls

File: src/tokens/ERC1155.sol (hunk 3 above)
Comment (transmissions11) [reply src/tokens/ERC1155.sol:134]: nw nw idrc if its draft or not lol

**Comment by abigger87** [comment]
> also no natspec or comments unless behavior super confusing pls

😢

File: src/tokens/ERC1155.sol [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  function _safeTransferFrom(address from, address to, uint256 id, uint256 amount) internal {
+    balances[from][id] = balances[from][id] - amount;
+    balances[to][id] = balances[to][id] + amount;
+
+    emit TransferSingle(msg.sender, from, to, id, amount);
+  }
+
+  function _callonERC1155Received(address from, address to, uint256 id, uint256 amount, uint256 gasLimit, bytes memory data) internal {
+    if (to.code.length > 0) {
+      bytes4 retval = ERC1155TokenReceiver(to).onERC1155Received{gas: gasLimit}(msg.sender, from, id, amount, data);
```
Comment (abigger87) [inline src/tokens/ERC1155.sol:66]: What should I do here to use the onERC1155Received param on `to`?

File: src/tokens/ERC1155.sol [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+      balances[from][ids[i]] = balances[from][ids[i]] - amounts[i];
+      balances[to][ids[i]] = balances[to][ids[i]] + amounts[i];
+    }
+
+    emit TransferBatch(msg.sender, from, to, ids, amounts);
+  }
+
+  function _callonERC1155BatchReceived(address from, address to, uint256[] memory ids, uint256[] memory amounts, uint256 gasLimit, bytes memory data) internal {
+    if (to.code.length > 0) {
+      bytes4 retval = ERC1155TokenReceiver(to).onERC1155BatchReceived{gas: gasLimit}(msg.sender, from, ids, amounts, data);
```
Comment (abigger87) [inline src/tokens/ERC1155.sol:85]: ++

File: src/tokens/ERC1155.sol (hunk 5 above)
Comment (transmissions11) [reply src/tokens/ERC1155.sol:66]: that looks fine— what am i missing?

File: src/tokens/ERC1155.sol [hunk 7]
```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity >=0.8.0;
+
+import {ERC1155TokenReceiver} from "./ERC1155TokenReceiver.sol";
+
+/// @notice Modern and gas efficient ERC1155 implementation.
+/// @author Modified from 0xsequence (https://github.com/0xsequence/erc-1155)
+abstract contract ERC1155 is ERC1155TokenReceiver {
```
Comment (abigger87) [inline src/tokens/ERC1155.sol:8]: Should this also be an ERC1155TokenReceiver @transmissions11 ? Don't have default impls in the ERC1155TokenReceiver as well, should I add?

Comment (transmissions11) [reply src/tokens/ERC1155.sol:8]: wait why would it be a reciver? the contract shouldnt accept its on tokens right?

Comment (transmissions11) [reply src/tokens/ERC1155.sol:8]: in terms of default impls nah lets just keep an interface

**Comment by transmissions11** [comment]
thanks king will clean up from here

**Pull request merged**